File: Cpp Codes/Motion_blur.cpp
#include "highgui.h"
#include "cv.h"
#include "string"
#include "math.h"
#include "iostream"
#include "limits"

using namespace::std;
using namespace::cv;

#define PI 3.14159265358979323846
float EPS = numeric_limits<float>::epsilon();

Mat Motion_blur(double len, double angle)
{
	double half = len / 2;
	double alpha = (angle - floor(angle / 180) * 180) / 180 * PI;
	double cosalpha = cos(alpha);
	double sinalpha = sin(alpha);
	int xsign;
	if (cosalpha < 0)
	{
		xsign = -1;
	}
	else
	{
		if (angle == 90)
		{
			xsign = 0;
		}
		else
		{
			xsign = 1;
		}
	}
	int psfwdt = 1;
	int sx = (int)fabs(half*cosalpha + psfwdt*xsign - len*EPS);
	int sy = (int)fabs(half*sinalpha + psfwdt - len*EPS);
	Mat_<double> psf1(sy, sx, CV_64F);
	Mat_<double> psf2(sy * 2, sx * 2, CV_64F);
	int row = 2 * sy;
	int col = 2 * sx;
	/*为减小运算量，先计算一半大小的PSF*/
	for (int i = 0; i < sy; i++)
	{
		double* pvalue = psf1.ptr<double>(i);
		for (int j = 0; j < sx; j++)
		{
			pvalue[j] = i*fabs(cosalpha) - j*sinalpha;
				double rad = sqrt(i*i + j*j);
			if (rad >= half && fabs(pvalue[j]) <= psfwdt)
			{
				double temp = half - fabs((j + pvalue[j] * sinalpha) / cosalpha);
				pvalue[j] = sqrt(pvalue[j] * pvalue[j] + temp*temp);
			}
			pvalue[j] = psfwdt + EPS - fabs(pvalue[j]);
			if (pvalue[j] < 0)
			{
				pvalue[j] = 0;
			}
		}
	}
	/*将模糊核矩阵扩展至实际大小*/
	for (int i = 0; i < sy; i++)
	{
		double* pvalue1 = psf1.ptr<double>(i);
		double* pvalue2 = psf2.ptr<double>(i);
		for (int j = 0; j < sx; j++)
		{
			pvalue2[j] = pvalue1[j];
		}
	}
		for (int i = 0; i < sy; i++)
	{
		for (int j = 0; j < sx; j++)
		{
			psf2[2 * sy - 1 - i][2 * sx - 1 - j] = psf1[i][j];
			psf2[sy + i][j] = 0;
			psf2[i][sx + j] = 0;
		}
	}
	/*保持图像总能量不变，归一化矩阵*/
	double sum = 0;
	for (int i = 0; i < row; i++)
	{
		for (int j = 0; j < col; j++)
		{
			sum += psf2[i][j];
		}
	}
	psf2 = psf2 / sum;
	if (cosalpha>0)
	{
		flip(psf2, psf2, 0);
	}
		//cout << "psf2=" << psf2 << endl;

	return psf2;

}

int main()
{
	int length = 100;
	int angle = 89.999;
	Mat blur;

	Mat img = imread("D:/Desktop/WriteData/image/1.jpg");
	cvNamedWindow("Origin", CV_WINDOW_AUTOSIZE);
	cvNamedWindow("Blur", CV_WINDOW_AUTOSIZE);
	imshow("Origin", img);

	Mat psf = Motion_blur(length, angle);
	filter2D(img, blur, -1, psf);

	imshow("Blur", blur);
	imwrite("D:/Desktop/WriteData/image/result.jpg", blur);
	cvWaitKey(0);
	cvDestroyWindow("From");
	cvDestroyWindow("To");
	return 0;
}

File: CV/motion_blur.py
import cv2
import numpy as np
import math

def generate_kernel():
    size = 15
    # generating the kernel
    kernel_motion_blur = np.zeros((size, size))
    kernel_motion_blur[int((size-1)/2), :] = np.ones(size)
    kernel = kernel_motion_blur / size

    return kernel

def genarate_Psf(length,angle):
    EPS=np.finfo(float).eps                                 
    alpha = (angle-math.floor(angle/ 180) *180) /180* math.pi
    cosalpha = math.cos(alpha)  
    sinalpha = math.sin(alpha)
    half = length / 2
    if cosalpha < 0:
        xsign = -1
    elif angle == 90:
        xsign = 0
    else:  
        xsign = 1
    psfwdt = 1;  

    sx = int(math.fabs(length*cosalpha + psfwdt*xsign - length*EPS))  
    sy = int(math.fabs(length*sinalpha + psfwdt - length*EPS))
    psf1=np.zeros((sy,sx))
     

    for i in range(0,sy):
        for j in range(0,sx):
            psf1[i][j] = i*math.fabs(cosalpha) - j*sinalpha
            rad = math.sqrt(i*i + j*j) 
            if  rad >= half and math.fabs(psf1[i][j]) <= psfwdt:  
                temp = half - math.fabs((j + psf1[i][j] * sinalpha) / cosalpha)  
                psf1[i][j] = math.sqrt(psf1[i][j] * psf1[i][j] + temp*temp)
            psf1[i][j] = psfwdt + EPS - math.fabs(psf1[i][j]);  
            if psf1[i][j] < 0:
                psf1[i][j] = 0

    anchor=(0,0)

    if angle<90 and angle>0:
        psf1=np.fliplr(psf1)
        anchor=(psf1.shape[1]-1,0)
    elif angle>-90 and angle<0:
        psf1=np.flipud(psf1)
        psf1=np.fliplr(psf1)
        anchor=(psf1.shape[1]-1,psf1.shape[0]-1)
    elif anchor<-90:
        psf1=np.flipud(psf1)
        anchor=(0,psf1.shape[0]-1)
    
    psf1=psf1/psf1.sum()
    
    return psf1,anchor

    
def main(argv=None):
    kernel, anchor = genarate_Psf(50, 0.1)
    print(kernel)
    img = cv2.imread('D:/Desktop/WriteData/image/2.jpg')
    cv2.imshow('Original', img)
    output = cv2.filter2D(img, -1, kernel, anchor=anchor)
    cv2.imshow('Motion Blur', output)
    cv2.imwrite("D:/Desktop/WriteData/image/blur2.jpg", output)
    cv2.waitKey(0)

if __name__ == '__main__':
    main()
